broadinstitute/single_cell_portal_core
# Selection UI for author-computed differential expression (SCP-4925)

This lets users select DE computed by authors, enabling new types of comparison on a broader set of annotations.

### Overview
Previously, DE was only available for cell-type-like groups, as one-vs-rest comparisons, and it was computed by SCP.  This is  popular with users when it's available, but still rather limited.

Now, study owners or editors ("authors" in viewer-facing text) can upload DE results that they have precomputed and refined.  These results can include only one-vs-rest comparisons like before, or only pairwise comparisons, or both.  They can also include DE results not only for just cell-type-like annotations, but any group-based annotation with > 1 group.

This author-computed DE also enables study owners to upload data that is non-exhaustive.  In other words, owners can upload DE for an incomplete set of comparisons.

These updates also make the DE dot plot call-to-action more salient.

### Scenarios
Author-computed DE enables the following the comparison profiles for each annotation group.  Here "all" means "for all potential comparisons in the annotation in this comparison type", and similarly for "some" and "none".

1. All one-vs-rest, all pairwise
2. All one-vs-rest, some pairwise
3. All one-vs-rest, no pairwise
4. Some one-vs-rest, all pairwise
4. No one-vs-rest, all pairwise
5. Some one-vs-rest, some pairwise
6. No one-vs-rest, no pairwise

### Video
Here's how it looks.

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/22e003ef-30bf-4769-b132-ebc6ea530860

### Test
An automated test covers basic functionality.  If desired, to manually test:

1.  Go to "Human milk - differential expression" study.  If on staging after this is merged, that's [SCP303](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP303/human-milk-differential-expression).
2. Select annotation "General_Celltype"
3. Click "Differential expression"
4. Confirm two select drop-down menus appear
5. Select "B cells"
6. Click into 2nd menu, confirm it says "rest" then does _not_ say "CSN1S1 macrophages", then says "dendritic cells".  This verifies handling of incomplete pairwise comparisons.
7. Select "rest" in 2nd menu
8. Confirm DE table displays with content in rows
9. Confirm there is a button that says "Dot plot" with an icon at left of the text
10. In the 1st menu, select "CSN1S1 macrophages"
11. Confirm the DE table is removed from the DE panel
12. Click into the 2nd menu, confirm it does _not_ include "rest".  This verifies handling of incomplete one-vs-rest comparisons.  Also confirm it does not include "B cells", which verifies handling of naturally-sorted incomplete pairwise comparisons.
13. Open DevTools, go to Network panel
14. Select "dendritic cells" in 2nd menu
15. Note that no content loads in DE panel, which is expected
16. In Network panel, confirm a request is sent for a URL that includes "cluster_umap_txt--General_Celltype--CSN1S1_macrophages--dendritic_cells--study--wilcoxon.tsv".

Loading author-computed DE results into the DE table will be handling in separate work within the next two weeks.

### Further detail
More context and discussion is available in [this SCP demo video](https://drive.google.com/file/d/1t6oeMQnjgRiWUKkezzffy24_I4k288xC/view?t=3m24s).  

This satisfies SCP-4925.
